src/compiler/checker.ts

```diff
@@ -42782,13 +42816,12 @@ namespace ts {
             }
 
             const nodeArguments = node.arguments;
-            if (nodeArguments.length !== 1) {
-                return grammarErrorOnNode(node, Diagnostics.Dynamic_import_must_have_one_specifier_as_an_argument);
+            if (checkGrammarImportCallArguments(node, nodeArguments)) {
+                return true;
             }
-            checkGrammarForDisallowedTrailingComma(nodeArguments);
             // see: parseArgumentOrArrayLiteralElement...we use this function which parse arguments of callExpression to parse specifier for dynamic import.
             // parseArgumentOrArrayLiteralElement allows spread element to be in an argument list which is not allowed as specifier in dynamic import.
-            if (isSpreadElement(nodeArguments[0])) {
+            if (nodeArguments.length && isSpreadElement(nodeArguments[0])) {

```

Comment:
Do we need to check that all elements of the `assert` property of the second argument have type `string`? Per https://tc39.es/proposal-import-assertions/#sec-evaluate-import-call, Step 10.d.v.3:

![image](https://user-images.githubusercontent.com/3902892/132757135-a420c1c6-2211-42b0-8429-4478a26174f4.png)

No coercion to string is performed here, the algorithm explicitly checks for strings.

We might want to consider something like the `ImportMeta` interface for the import argument, something like:

```ts
// es5.d.ts
interface ImportCallOptions {
  assert?: ImportAssertions;
}
interface ImportAssertions {
  type?: string;
  [key: string]: string | undefined; // not ideal, but necessary so that `type` is optional.
}
```
Then we can check against the global `ImportCallOptions` type, and if other assertions or options are added later, they can be introduced using global augmentation similar to `ImportMeta`.

We could consider this as a follow-on PR if necessary, but it would be nice to have.

Reply:
NOTE: if we wanted to be *more* specific with `type` we'd need to create an open-ended union via an interface as well:

```ts
interface ImportAssertions {
  type?: ImportTypeAssertion;
  [key: string]: string | undefined;
}

// see ArrayBufferTypes for reference
interface ImportTypeAssertionTypes {
  json: "json";
}

type ImportTypeAssertion = Extract<ImportTypeAssertions[keyof ImportTypeAssertions], string>;
```

That way, we have better completions for `type` and if a host environment supported other values for `type`, they could also be added via augmentation, i.e.:

```ts
interface ImportTypeAssertionTypes {
  css: "css";
}
```

Reply:
Also, the `nodeArguments.length` check shouldn't be necessary if `checkGrammarImportCallArguments` already asserts that `nodeArguments.length` can't be zero.

Reply:
We also need to ensure that `nodeArguments[1]` (if present) is not a spread element.

I would suggest this approach:
```ts
const spreadElement = forEach(nodeArguments, isSpreadElement);
if (spreadElement) {
  return grammarErrorOnNode(spreadElement, Diagnostics.Specifier_of_dynamic_import_cannot_be_spread_element);
}
```

That way it checks all arguments, but only reports on the first error (which is almost always the case for grammar errors).

Reply:
Should we check the assert clause too？Or just the second parameter of dynamic import.